### PR TITLE
Sort by popular

### PR DESCRIPTION
**Pull request by Ru-T**
## What Changed? 
* Added field and counter_cache for line_items belonging to books
* Added sort param in index action of books controller
* Added most popular sort link to books index view
* Pass feature spec for sorting by popular

## Why? 
* So users can sort by most popular books and buy em!

### CODE REVIEW

File: app/admin/book.rb [hunk 1]
```diff
@@ -4,11 +4,11 @@
...
-    column :price
+    column :price_cents
     column :category
     column :description
     actions
   end
 
-  permit_params :title, :author, :published_date, :price, :category, :description
+  permit_params :title, :author, :published_date, :price_cents, :category, :description
 end
```
Comment (Ru-T) [inline app/admin/book.rb]: Fixed a failing admin spec caused by last merge. Unrelated to this PR's purpose though.

File: features/step_definitions/administration_steps.rb [hunk 2]
```diff
@@ -50,7 +50,7 @@
 end
 
 When(/^I enter the price "(.*?)"$/) do |price|
-  fill_in "Price", with: price
+  fill_in "Price cents", with: price
 end
 
 And(/^I select the date "([^"]*)" for "(.*?)"$/) do |in_date_str, field|
```
Comment (Ru-T) [inline features/step_definitions/administration_steps.rb]: Related to above unrelated fix.

File: features/step_definitions/authentication_steps.rb [hunk 3]
```diff
@@ -8,9 +8,8 @@
 end
 
 Given(/^My account is confirmed$/) do
-  ctoken = ActionMailer::Base.deliveries.last.body.match(/confirmation_token=\w*/)
-  visit "/users/confirmation?#{ctoken}"
   user = User.find_for_authentication(email: 'newestuser@test.com')
+  user.confirmed_at = Date.today
   expect(user).to be_confirmed
 end
```
Comment (Ru-T) [inline features/step_definitions/authentication_steps.rb]: Another failing spec (only periodically) but fixed it here. Unrelated to PR's purpose.

File: app/admin/book.rb [hunk 4]
```diff
@@ -4,11 +4,11 @@
     column :title
     column :author
     column :published_date
-    column :price
+    column :price_cents
```
Comment (dkniffin) [inline app/admin/book.rb]: You may want to change this. It's more user-friendly to show this column in dollars.

File: app/controllers/books_controller.rb [hunk 5]
```diff
@@ -3,8 +3,10 @@ class BooksController < ApplicationController
   before_action :set_book, only: [:show]
 
   def index
-    @q = Book.ransack(params[:q])
-    @books = @q.result(distinct: true).page(params[:page])
+    sort = params[:sort] || "published_date DESC"
+    sort = "order_count DESC" if sort == "most_popular"
```
Comment (dkniffin) [inline app/controllers/books_controller.rb]: You could remove this line if you change [this line](https://github.com/smashingboxes/ruti-bookstore/pull/27/files#diff-b9cabdc15aa47777492ecb56db9bfac3R5) to be `sort: "order_count"`

File: features/step_definitions/authentication_steps.rb (hunk 3 above)
Comment (dkniffin) [inline features/step_definitions/authentication_steps.rb]: :+1:

File: features/step_definitions/book_index_page_steps.rb [hunk 6]
```diff
@@ -1,5 +1,7 @@
 Given(/^there are 100 books in the database$/) do
-  FactoryGirl.create_list(:book, 100)
+  FactoryGirl.create(:book, published_date: Date.today, order_count: 2)
+  FactoryGirl.create(:book, order_count: 1)
+  FactoryGirl.create_list(:book, 98)
 end
```
Comment (dkniffin) [inline features/step_definitions/book_index_page_steps.rb]: One improvement I would make here is to take the `100` in as an argument, so if you had to do `Given there are 10 books in the database` you wouldn't have to write a new step.

File: features/step_definitions/book_index_page_steps.rb [hunk 7]
```diff
@@ -28,3 +29,16 @@
 Then(/^I am shown a list of books with that title$/) do
   expect(page).to have_content(Book.find(1).title)
 end
+
+Given(/^some books have been ordered$/) do
+  @books = Book.order(published_date: :desc)
```
Comment (dkniffin) [inline features/step_definitions/book_index_page_steps.rb]: I think you mis-understand "ordered" here. I think it means purchased. You're doing that above in the `there are 100 books in the database`, but I think it should go in here.

File: features/step_definitions/book_index_page_steps.rb [hunk 8]
```diff
@@ -7,8 +9,7 @@
 end
 
 Then(/^the books are ordered by published date$/) do
-  expect(Book.first).to eq Book.order(:published_date).first
-  expect(Book.last).to eq Book.order(:published_date).last
+  expect(Book.first).to eq Book.order(published_date: :desc).first
```
Comment (dkniffin) [inline features/step_definitions/book_index_page_steps.rb]: Here, this is just checking the database. You're not actually checking anything on the page.

File: features/step_definitions/book_index_page_steps.rb [hunk 9]
```diff
@@ -28,3 +29,16 @@
...
+  @books = Book.order(published_date: :desc)
+end
+
+When(/^I sort by "([^"]*)"$/) do |sort|
+  click_on sort
+end
+
+Then(/^the books are re\-sorted based on the amount of times they are purchased$/) do
+  Book.order(order_count: :desc)
+  expect(Book.first).to eq Book.order(order_count: :desc).first
```
Comment (dkniffin) [inline features/step_definitions/book_index_page_steps.rb]: Same here. You're not checking anything on the page.

File: app/admin/book.rb (hunk 4 above)
Comment (Ru-T) [inline app/admin/book.rb]: @dkniffin Will do in my next PR which is admin focused. Cool?

Comment (dkniffin) [inline app/admin/book.rb]: :+1:

File: app/controllers/books_controller.rb (hunk 5 above)
Comment (Ru-T) [inline app/controllers/books_controller.rb]: Updated.

File: features/step_definitions/book_index_page_steps.rb (hunk 6 above)
Comment (Ru-T) [inline features/step_definitions/book_index_page_steps.rb]: :+1:

File: features/step_definitions/book_index_page_steps.rb (hunk 7 above)
Comment (Ru-T) [inline features/step_definitions/book_index_page_steps.rb]: haha whoops. ordered has too many meanings!

Comment (dkniffin) [inline features/step_definitions/book_index_page_steps.rb]: Yea. It might be a good idea to update the feature file. to say "some books have been purchased". Wanna submit a PR for that?

File: features/step_definitions/book_index_page_steps.rb (hunk 9 above)
Comment (Ru-T) [inline features/step_definitions/book_index_page_steps.rb]: Updated.

**Comment by Ru-T** [comment]
@dkniffin All comments integrated! Let me know what you think.

**Comment by Ru-T** [comment]
@dkniffin Fingers crossed, should be good now! :)

File: features/step_definitions/book_index_page_steps.rb [hunk 10]
```diff
@@ -7,8 +7,7 @@
 end
 
 Then(/^the books are ordered by published date$/) do
-  expect(Book.first).to eq Book.order(:published_date).first
-  expect(Book.last).to eq Book.order(:published_date).last
+  expect(page).to have_content(/#{Book.first.title}.*#{Book.fifth.title}/m)
```
Comment (dkniffin) [inline features/step_definitions/book_index_page_steps.rb]: This looks good now. The fact that there's a `fifth` method still baffles me.

**Comment by dkniffin** [comment]
Yep, looks good! :shipit:
